### ISSUE TITLE
background fill stops working on adding type and scale to axis

### ISSUE THREAD

**Issue by shrutikaponde** [recharts/recharts#1787]
background color do not change on hover, when type and scale props are given to XAxis

```
<XAxis
            dataKey="value"
            type="number" 
            scale="time"  
            domain={['dataMin', 'dataMax']}
            tickFormatter={(tick) => moment(tick * 1000).format('MMM Do')}
            interval='preserveStartEnd'
            ticks={[1527811200,1538352000,1548979200,1559347200]}
 />
```

![no_fill](https://user-images.githubusercontent.com/26830541/60439691-d9ec0e00-9c30-11e9-9e31-d1cc00f1c090.png)

background color changes on hover, but Xaxis is not visible, on commenting type and scale.
```
 <XAxis
            dataKey="value"
            // type="number" 
            // scale="time"
            domain={['dataMin', 'dataMax']}
            tickFormatter={(tick) => moment(tick * 1000).format('MMM Do')}
            interval='preserveStartEnd'
            ticks={[1527811200,1538352000,1548979200,1559347200]}
    />
```
![no-axis](https://user-images.githubusercontent.com/26830541/60439833-1f104000-9c31-11e9-8188-b05700feb181.png)

On commenting type, scale and ticks, background color changes but i want to use ticks.
```
 <XAxis
            dataKey="value"
            // type="number" 
            // scale="time"
            domain={['dataMin', 'dataMax']}
            tickFormatter={(tick) => moment(tick * 1000).format('MMM Do')}
            interval='preserveStartEnd'
            // ticks={[1527811200,1538352000,1548979200,1559347200]}
       />
```
![fill](https://user-images.githubusercontent.com/26830541/60440012-7f06e680-9c31-11e9-9545-57ac42128cbc.png)

how can i get background color change with ticks on XAxis ?

**Comment by ice1080** [comment]
Here is a more full example, which shows that it's just `type='number'` that breaks the hover and scale is not even needed:

```
<ResponsiveContainer width={'100%'} height={'100%'}>
    <BarChart data={[{test: 1, number: 1}, {test: 2, number: 2}, {test: 3, number: 3}]}>
        <XAxis dataKey={'number'} type={'number'}/>
        <YAxis/>
        <Tooltip cursor={{fill: 'yellow'}}/>
        <Bar dataKey={'test'}/>
    </BarChart>
</ResponsiveContainer>
```

Changing the above to just `<XAxis>` allows the hover text to work again, but removes the ability to use domain (or scale as mentioned above).

Comfirmed this is still an issue on version 2.0.9 (I'm using react 17.0.2).

**Comment by ice1080** [comment]
Digging into the code for this a bit, though I'm not great with ts so it's slow going.
It looks like Rectangle is receiving a width of 0 from generateCategoricalChart.renderCursor. 
Going further, it looks like getBandSizeOfAxis in ChartUtils does not have a scale to get the bandwidth from when type = number. So either that needs to get set or we need a different way of handling things when the user is user an XAxis with type of number and not supplying a list of ticks. In my specific use case I'm supplying a tickCount and a domain. I'll try to keep looking into this as I have time.

**Comment by ice1080** [comment]
In ChartUtils.parseScale, for a scale of 'auto' it looks like when type is category (default) it uses d3Scales.scaleBand(), and when type is number it uses d3Scales.scaleLinear(). I'm guessing that changing that would cause many other problems. 
However, after some more digging I noticed that this is the only instance of getBandSizeOfAxis that only passes in the axis and doesn't also pass the ticks. Since we already had the tooltipTicks in tooltipTicksGenerator, I tried passing those in and saw this working. Unit tests pass so I will fork and create a PR for this shortly.
